db/hybrid_logical_vector.go

```diff
@@ -180,6 +193,43 @@ func (hlv *HybridLogicalVector) GetVersion(sourceID string) uint64 {
 	return latestVersion
 }
 
+// AddNewerVersions will take a hlv and add any newer source/version pairs found across CV and PV found in the other HLV taken as parameter
+func (hlv *HybridLogicalVector) AddNewerVersions(otherVector HybridLogicalVector) {
+	if otherVector.PreviousVersions != nil || len(otherVector.PreviousVersions) != 0 {
+		// Iterate through incoming vector previous versions, update with the version from other vector
+		// for source if the local version for that source is lower
+		for i, v := range otherVector.PreviousVersions {
+			if hlv.PreviousVersions[i] == 0 || hlv.PreviousVersions[i] < v {
+				hlv.setPreviousVersion(i, v)
+			}
+		}
+	}
+
+	// create current version for incoming vector and attempt to add it to the local HLV, AddVersion will handle if attempting to add older
+	// version than local HLVs CV pair
+	otherVectorCV := CurrentVersionVector{SourceID: otherVector.SourceID, VersionCAS: otherVector.Version}
+	err := hlv.AddVersion(otherVectorCV)
+	if err != nil {
+		// we get here if the other HLV has lower current version than local HLV, so we
+		// need to check if the current source between the two are not equal then the
+		// incoming CV pair needs to be added to PV on local HLV
+		if hlv.SourceID != otherVector.SourceID {
+			if hlv.PreviousVersions[otherVector.SourceID] == 0 {
+				// source doesn't exist in pv so add it this pair
+				hlv.setPreviousVersion(otherVector.SourceID, otherVector.Version)
+			} else if hlv.PreviousVersions[otherVector.SourceID] != 0 && hlv.PreviousVersions[otherVectorCV.SourceID] < otherVector.Version {
+				// source exists but has version less than incoing version associated with it so replace this value in PV
+				// with incoming pair
+				hlv.setPreviousVersion(otherVector.SourceID, otherVector.Version)
+			}
+		}
+	}
+	// if current source exists in PV, delete it.
+	if hlv.PreviousVersions[hlv.SourceID] != 0 {
```

Comment:
Same comment here about using existence check instead of ascribing meaning to version=0.